### PR TITLE
KTOR-1661 Fix early dispatcher close

### PR DESCRIPTION
**Pull request by e5l**
Fix [KTOR-1661](https://youtrack.jetbrains.com/issue/KTOR-1661)

### CODE REVIEW

File: ktor-client/ktor-client-java/jvm/src/io/ktor/client/engine/java/JavaHttpEngine.kt [hunk 1]
```diff
@@ -45,7 +45,17 @@ public class JavaHttpEngine(override val config: JavaHttpConfig) : HttpClientEng
 
     private var httpClient: HttpClient? = null
 
+    override val coroutineContext: CoroutineContext
+    private val requestsJob: CoroutineContext
```
Comment (rsinukov) [inline ktor-client/ktor-client-java/jvm/src/io/ktor/client/engine/java/JavaHttpEngine.kt:49]: can be local variable

File: ktor-client/ktor-client-features/ktor-client-websockets/common/test/io/ktor/client/tests/features/WebSocketRemoteTest.kt [hunk 2]
```diff
@@ -13,7 +13,7 @@ import kotlinx.coroutines.*
 import kotlin.test.*
 
 class WebSocketRemoteTest : ClientLoader() {
-    private val echoWebsocket = "echo.websocket.org"
+    private val echoWebsocket = "$TEST_WEBSOCKET_SERVER/websockets/echo"
```
Comment (rsinukov) [inline ktor-client/ktor-client-features/ktor-client-websockets/common/test/io/ktor/client/tests/features/WebSocketRemoteTest.kt:16]: is it related?

**Review by rsinukov — APPROVED** [review]
LGTM, but minor comments

File: ktor-client/ktor-client-features/ktor-client-websockets/common/test/io/ktor/client/tests/features/WebSocketRemoteTest.kt (hunk 2 above)
Comment (e5l) [reply ktor-client/ktor-client-features/ktor-client-websockets/common/test/io/ktor/client/tests/features/WebSocketRemoteTest.kt:16]: Can extract in the separate PR if needed

File: ktor-client/ktor-client-java/jvm/src/io/ktor/client/engine/java/JavaHttpEngine.kt (hunk 1 above)
Comment (e5l) [reply ktor-client/ktor-client-java/jvm/src/io/ktor/client/engine/java/JavaHttpEngine.kt]: can be removed

File: ktor-client/ktor-client-java/jvm/src/io/ktor/client/engine/java/JavaHttpEngine.kt [hunk 3]
```diff
@@ -45,7 +45,14 @@ public class JavaHttpEngine(override val config: JavaHttpConfig) : HttpClientEng
 
     private var httpClient: HttpClient? = null
 
+    override val coroutineContext: CoroutineContext
+
     init {
+        val parentContext = super.coroutineContext
+        val parent = parentContext[Job]!!
```
Comment (cy6erGn0m) [inline ktor-client/ktor-client-java/jvm/src/io/ktor/client/engine/java/JavaHttpEngine.kt:52]: So now we have two jobs and only one of them is completed. Who will complete the `parent`?

File: ktor-client/ktor-client-core/jvm/test/HttpClientEngineTest.kt [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+ */
+
+import io.ktor.client.*
+import io.ktor.client.utils.*
+import io.ktor.client.engine.mock.*
+import io.ktor.test.dispatcher.*
+import kotlinx.coroutines.*
+import kotlin.test.*
+
+/*
```
Comment (cy6erGn0m) [inline ktor-client/ktor-client-core/jvm/test/HttpClientEngineTest.kt:12]: doble license

**Review by cy6erGn0m — APPROVED** [review]

File: ktor-client/ktor-client-core/jvm/test/HttpClientEngineTest.kt (hunk 4 above)
Comment (e5l) [reply ktor-client/ktor-client-core/jvm/test/HttpClientEngineTest.kt]: thanks,  fixed

File: ktor-client/ktor-client-java/jvm/src/io/ktor/client/engine/java/JavaHttpEngine.kt (hunk 3 above)
Comment (e5l) [reply ktor-client/ktor-client-java/jvm/src/io/ktor/client/engine/java/JavaHttpEngine.kt]: reverted, will investigate it for other engines

**Pull request merged**